plugin/storage/badger/spanstore/reader.go

```diff
@@ -494,34 +513,34 @@ func validateQuery(p *spanstore.TraceQueryParameters) error {
 }
 
 // scanIndexKeys scans the time range for index keys matching the given prefix.
-func (r *TraceReader) scanIndexKeys(indexKeyValue []byte, startTimeMin time.Time, startTimeMax time.Time) ([][]byte, error) {
+func (r *TraceReader) scanIndexKeys(indexKeyValue []byte, plan *executionPlan) ([][]byte, error) {
 	indexResults := make([][]byte, 0)
 
-	startStampBytes := make([]byte, 8)
-	binary.BigEndian.PutUint64(startStampBytes, model.TimeAsEpochMicroseconds(startTimeMin))
-
 	err := r.store.View(func(txn *badger.Txn) error {
 		opts := badger.DefaultIteratorOptions
 		opts.PrefetchValues = false // Don't fetch values since we're only interested in the keys
+		opts.Reverse = true
 		it := txn.NewIterator(opts)
 		defer it.Close()
 
 		// Create starting point for sorted index scan
-		startIndex := make([]byte, len(indexKeyValue)+len(startStampBytes))
+		startIndex := make([]byte, len(indexKeyValue)+8+1)
+		startIndex[len(startIndex)-1] = 0xFF
 		copy(startIndex, indexKeyValue)
-		copy(startIndex[len(indexKeyValue):], startStampBytes)
+		copy(startIndex[len(indexKeyValue):], plan.startTimeMax)
 
-		timeMax := model.TimeAsEpochMicroseconds(startTimeMax)
-		for it.Seek(startIndex); scanFunction(it, indexKeyValue, timeMax); it.Next() {
+		for it.Seek(startIndex); scanFunction(it, indexKeyValue, plan.startTimeMin); it.Next() {
 			item := it.Item()
 
 			// ScanFunction is a prefix scanning (since we could have for example service1 & service12)
 			// Now we need to match only the exact key if we want to add it
 			timestampStartIndex := len(it.Item().Key()) - (sizeOfTraceID + 8) // timestamp is stored with 8 bytes
 			if bytes.Equal(indexKeyValue, it.Item().Key()[:timestampStartIndex]) {
-				key := make([]byte, len(item.Key()))
-				copy(key, item.Key())
-				indexResults = append(indexResults, key)
+				traceIDBytes := item.Key()[len(item.Key())-sizeOfTraceID:]
```

Comment:
Wondering whether it would be better to move knowledge about how to extract `traceIDBytes` from the `key` into the `Item` type - or possibly have a `Key` type that encapsulates all the necessary knowledge about how the key is encoded? Could be done as a separate PR, but might make it easier to maintain.

Reply:
The key is encoded in different way depending on the type of key. For example, primary keys that store the trace have different structure than the indexKeys.

Also, Item/Key are types used internally by badger, so that might be a bit confusing if those names were used. But yes, this could be built as "indexKey" and "traceKey" structs.